### PR TITLE
MGMT-8809: Fix e2e-metal-assisted-operator-disconnected CI job

### PR DESCRIPTION
**Pull request by javipolo**
In this PR we try to address are multiple problems when running e2e-metal-assisted-operator-disconnected CI job
Most of the problem arise only while running in disconnected mode:
- The assisted service operator catalog name is different while disconnected, and this needs to be consistent throughout the code
- If OS_IMAGES has multiple entries, all of them need to be mirrored
- While handling multiple mirrored artifacts we need to keep their full url path on the mirror to prevent name collisions
- Mirrored OS_IMAGES have to be added to AgentServiceConfig to override the default ones
- Due to a bug in hive, the condition that checks that ClusterDeployment is installed needs longer wait time

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [x] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by mkowalski** [comment]
```
[root@ipi-ci-op-1mn3q2ps-7d865-1501520675209220096 ~]# oc get nodes
NAME                   STATUS   ROLES           AGE   VERSION
ostest-extraworker-0   Ready    master,worker   32m   v1.21.6+4b61f94
[root@ipi-ci-op-1mn3q2ps-7d865-1501520675209220096 ~]# oc get clusterversion
NAME      VERSION   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.8.33    True        False         12m     Cluster version is 4.8.33
[root@ipi-ci-op-1mn3q2ps-7d865-1501520675209220096 ~]# oc get co
NAME                                       VERSION   AVAILABLE   PROGRESSING   DEGRADED   SINCE
authentication                             4.8.33    True        False         False      13m
baremetal                                  4.8.33    True        False         False      25m
cloud-credential                           4.8.33    True        False         False      29m
cluster-autoscaler                         4.8.33    True        False         False      25m
config-operator                            4.8.33    True        False         False      29m
console                                    4.8.33    True        False         False      13m
csi-snapshot-controller                    4.8.33    True        False         False      26m
dns                                        4.8.33    True        False         False      22m
etcd                                       4.8.33    True        False         False      26m
image-registry                             4.8.33    True        False         False      17m
ingress                                    4.8.33    True        False         False      20m
insights                                   4.8.33    True        False         False      26m
kube-apiserver                             4.8.33    True        False         False      21m
kube-controller-manager                    4.8.33    True        False         False      21m
kube-scheduler                             4.8.33    True        False         False      25m
kube-storage-version-migrator              4.8.33    True        False         False      29m
machine-api                                4.8.33    True        False         False      25m
machine-approver                           4.8.33    True        False         False      27m
machine-config                             4.8.33    True        False         False      20m
marketplace                                4.8.33    True        False         False      25m
monitoring                                 4.8.33    True        False         False      16m
network                                    4.8.33    True        False         False      30m
node-tuning                                4.8.33    True        False         False      26m
openshift-apiserver                        4.8.33    True        False         False      20m
openshift-controller-manager               4.8.33    True        False         False      20m
openshift-samples                          4.8.33    True        False         False      18m
operator-lifecycle-manager                 4.8.33    True        False         False      26m
operator-lifecycle-manager-catalog         4.8.33    True        False         False      26m
operator-lifecycle-manager-packageserver   4.8.33    True        False         False      22m
service-ca                                 4.8.33    True        False         False      29m
storage                                    4.8.33    True        False         False      26m
```

**Comment by mkowalski** [comment]
/approve

Can't LGTM because I took part in the work, we'll need another pair of eyes anyway

**Comment by osherdp** [comment]
@javipolo PR description has anything except what I need to understand the change :)
can you add some description around the issues with this job?

File: deploy/operator/gather.sh
Comment (osherdp) [inline deploy/operator/gather.sh:20]: We can, if we'll add ``source /root/config`` before running this script in [operator-gather step](https://github.com/openshift/release/blob/master/ci-operator/step-registry/assisted/baremetal/operator/gather/assisted-baremetal-operator-gather-commands.sh#L58)

Comment (osherdp) [reply deploy/operator/gather.sh:20]: I just don't want ``${LOGS_DEST}/${ASSISTED_SERVICE_OPERATOR_CATALOG}.log`` to get messy with an unrelated output like "could not find catalogsource ..."

File: deploy/operator/mirror_utils.sh
Comment (osherdp) [inline deploy/operator/mirror_utils.sh:94]: can you add a comment about how this parsing goes?

File: deploy/operator/setup_assisted_operator.sh
Comment (osherdp) [inline deploy/operator/setup_assisted_operator.sh:192]: those two lines are too complex I'm afraid
we need some way to make it more readable (even if longer in lines)

**Review by osherdp — CHANGES_REQUESTED** [review]

File: deploy/operator/mirror_utils.sh
Comment (javipolo) [reply deploy/operator/mirror_utils.sh:94]: This is in order to preserve the full path when mirroring the files. If not, when filenames are the same between different versions, they were overwritten

Extract the path from the URL by splitting by /, and keeping only fields 3 to end
```
$ echo https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.9/4.9.0/rhcos-live-rootfs.x86_64.img | cut -d / -f 3-
mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.9/4.9.0/rhcos-live-rootfs.x86_64.img
```

Get the directory name from the path using dirname
```
$ dirname mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.9/4.9.0/rhcos-live-rootfs.x86_64.img
mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.9/4.9.0
```

Running all linked together
```
$ dirname $(echo https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.9/4.9.0/rhcos-live-rootfs.x86_64.img | cut -d / -f 3-)
mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.9/4.9.0
```

Comment (osherdp) [reply deploy/operator/mirror_utils.sh:94]: ack, that's more understandable right now
find a way to add comments (``# bla bla``) or separate that to local variables so that it will be more immediate to get

File: deploy/operator/ztp/deploy_spoke_cluster.sh
Comment (mkowalski) [reply deploy/operator/ztp/deploy_spoke_cluster.sh:121]: And it’s there, ready to rebase

Comment (javipolo) [reply deploy/operator/ztp/deploy_spoke_cluster.sh]: Removed! :)

Comment (javipolo) [reply deploy/operator/ztp/deploy_spoke_cluster.sh]: Put it back (to 90s) because even if it's merged it seems it wont be released to use, so better safe than sorry for now

File: deploy/operator/utils.sh
Comment (osherdp) [inline deploy/operator/utils.sh:54]: Doesn't ``wait_for_object_amount`` do the same (and with timeout mechanism)?

**Review by osherdp — CHANGES_REQUESTED** [review]

File: deploy/operator/utils.sh
Comment (javipolo) [reply deploy/operator/utils.sh]: Oh, havent seen this function. You're right .... Changed!

**Review by osherdp — APPROVED** [review]

**Review by osherdp — APPROVED** [review]

**Comment by osherdp** [comment]
@javipolo from [e2e-metal-assisted-operator-disconnected](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-service/3445/pull-ci-openshift-assisted-service-master-e2e-metal-assisted-operator-disconnected/1501868480230592512):
```
export DISCONNECTED=false
```

that's weird, no?

**Comment by javipolo** [comment]
> ```
> export DISCONNECTED=false
> ```
> that's weird, no?

It's the same behaviour we were having before. I thought that my PR on `release` would fix it, but seems not :(

**Comment by osherdp** [comment]
/hold
waiting for the fix, for now we'll hold to save resources

**Comment by javipolo** [comment]
> @javipolo from [e2e-metal-assisted-operator-disconnected](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-service/3445/pull-ci-openshift-assisted-service-master-e2e-metal-assisted-operator-disconnected/1501868480230592512):
> ```
> export DISCONNECTED=false
> ```

@osherdp so far I could not find a way that this variable gets inherited by the gather job, so at this point we have 3 options:
1. Find out why this variable is not reaching gather.sh so this works properly
2. Rename the mirror-catalog-name to be the same one as the normal catalogsource name, so there are no discrepancies on naming
3. Try to guess the catalogname by checking if the mirror-catalog-name exists, but redirecting errors to /dev/null, so no "no catalogsource found" errors are thrown to the logs

I decided to go for number two, since name of the catalogsource is not important at all, and I don't even know why it was different name for starters

**Review by mkowalski — APPROVED** [review]

**Comment by osherdp** [comment]
@mkowalski in the meeting we talked about removing this job completely from 2.4 branch
so unless those changes are needed for another job, the rebase is not needed

**Comment by mkowalski** [comment]
I am not 100% sure... There are changes here about handling multiple entries in `os_images` which are not related to the disconnectness of this job. It may be that this part needs to be backported because (without checking) I'd expect that there are other ZTP jobs suffering from that. Anyway, let's give this one a moment to make CI more green and then, if we spot ZTP failing consistently in 2.4, we can backport what's needed

**Pull request merged**
